src/ledgerwriter/src/test/java/anthos/samples/financedemo/ledgerwriter/TransactionValidatorTest.java

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright 2020, Google LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package anthos.samples.financedemo.ledgerwriter;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+import org.mockito.Mock;
+
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.Mockito.when;
+import static org.mockito.MockitoAnnotations.initMocks;
+
+class TransactionValidatorTest {
+
+    private TransactionValidator transactionValidator;
+
+    @Mock
+    private Transaction transaction;
+
+    private static final String LOCAL_ROUTING_NUM = "123456789";
+    private static final String INVALID_NUM = "123456";
+    private static final String AUTHED_ACCOUNT_NUM = "1234567890";
+    private static final String NON_AUTHED_ACCOUNT_NUM = "0987654321";
+    private static final String TO_ACCOUNT_NUM = "5678901234";
+    private static final String TO_ROUTING_NUM = "567891234";
+    private static final Integer VALIDAMOUNT = 3755;
+    private static final Integer INVALIDAMOUNT = -23;
+
+    @BeforeEach
+    void setUp() {
+        initMocks(this);
+        transactionValidator = new TransactionValidator();
+
+        when(transaction.getFromAccountNum()).thenReturn(AUTHED_ACCOUNT_NUM);
+        when(transaction.getFromRoutingNum()).thenReturn(LOCAL_ROUTING_NUM);
+        when(transaction.getToAccountNum()).thenReturn(TO_ACCOUNT_NUM);
+        when(transaction.getToRoutingNum()).thenReturn(TO_ROUTING_NUM);
+        when(transaction.getAmount()).thenReturn(VALIDAMOUNT);
+    }
+
+    @Test
+    @DisplayName("Given the transaction is validated, no exception is thrown")
+    void validateTransactionSuccess() {
+        assertDoesNotThrow(() -> {
+            transactionValidator.validateTransaction(
+                    LOCAL_ROUTING_NUM, AUTHED_ACCOUNT_NUM, transaction);
+        });
+    }
+
+    @Test
+    @DisplayName("Given invalid sender account number, " +
+            "IllegalArgumentException is thrown")
+    void validateTransactionFailWhenInvalidFromAccountNumber() {
+        // Given
+        when(transaction.getFromAccountNum()).thenReturn(INVALID_NUM);
```

Comment:
Can we be more through here? It would be cool if we could test a list of a bunch of possible invalid values (8 digits, 10 digits, 9 characters, emojis, spaces, etc)

Reply:
Added a bunch of tests for validating account/routing numbers. Let me know if you think of more cases

Reply:
That's a lot of boilerplate code. Is it possible to just create an array of invalid values, and iterate through it in a loop?

Reply:
I couldn't decide because the boilerplate code has the benefit of having clear display name for each invalid number scenario. But I def agree that there's too much repetitive code. Added array and for loop to the test

Reply:
Looks good! I'd probably suggest removing the variables for all the invalid strings and just keeping the values in the arrays instead, but that's a nit

Reply:
Done!